### PR TITLE
Add SAC B path timeline migration

### PR DESCRIPTION
**Pull request by DH13768095744**
## Summary

This PR adds the SAC-only off-policy B path and makes it the default replay path for SAC training.

The existing A path remains available through:

```bash
training.replay_pipeline=gpu_cache
```

The new default behavior is:

```text
training.replay_pipeline=auto

SAC      -> cpu_pinned_double_buffer
TD3      -> gpu_cache
FlashSAC -> gpu_cache
```

This keeps TD3 and FlashSAC on the existing A path while making normal SAC training use the B path by default.

## What Changed

- Added a SAC-only `DoubleBufferOffPolicyRunner`.
- Added replay pipeline abstractions:
  - `gpu_cache` as the A path/control pipeline.
  - `cpu_pinned_double_buffer` as the B path implementation.
- Added the CPU-pinned double-buffer replay pipeline:
  - packed CPU replay storage;
  - collector-thread CPU packed sampling;
  - two shared host slots;
  - two GPU sampled-batch slots;
  - pybind11 H2D submit path.
- Added Perfetto/Chrome trace event recording for off-policy replay and learner timing.
- Added `scripts/analyze_offpolicy_trace.py` for timeline comparison.
- Updated SAC config dispatch so user-facing SAC training defaults to B path.
- Kept the old A path explicitly available with `training.replay_pipeline=gpu_cache`.
- Added focused tests for replay pipeline dispatch, B path constraints, trace markers, and analyzer behavior.

### A Path: `gpu_cache`

The original A path uses `OffPolicyRunner` with `ReplayBuffer.sample()`.

For CUDA, `ReplayBuffer` keeps the authoritative shared CPU replay tensors and also allocates a full GPU replay cache. On each learner sample, it lazily syncs newly written collector data into the GPU cache, then samples from GPU tensors with advanced indexing.

For the `G1WalkFlat / mujoco` SAC workload in the report:

```text
replay capacity = 1,048,576 transitions
packed width    = 430 float32 values
full GPU replay cache ~= 1720 MiB
lazy H2D per iter ~= 3.359 MiB
```

This path is still important as the explicit control path and remains available.

### B Path: `cpu_pinned_double_buffer`

The B path avoids the full GPU replay cache. It stores replay as packed CPU shared memory and prepares each learner iteration's sampled batch through a double-buffer pipeline.

The collector thread handles packed CPU sampling into a shared slot. The pybind11 path submits H2D into one of two GPU sampled-batch slots. The learner consumes the hot GPU slot while the next cold slot is prepared.

For the same workload:

```text
sample_count per learner iter = 32768 transitions
one sampled GPU slot ~= 53.75 MiB
two sampled GPU slots ~= 107.5 MiB
full GPU replay cache = not allocated
```

The result is not a pure speed win over A path. The value is near-A throughput while greatly reducing persistent replay GPU memory.

## Scope and Constraints

The B path is intentionally constrained in this PR:

- SAC only.
- Single GPU only.
- CUDA only.
- Synchronized collection only.
- `replay_prefetch_mode=one_tick`.
- Fixed internal B path implementation:
  - `pack_layout=packed`
  - `pack_executor=collector_thread`
  - `h2d_submitter=pybind11`
  - learner-light logging behavior fixed in code

The old experimental knobs are not exposed:

```text
replay_learner_lightweight
replay_learner_lightweight_log_interval
replay_pack_layout
replay_pack_executor
replay_h2d_submitter
```

## Performance Summary

The underlying report compared:

1. A path: `gpu_cache`
2. B current baseline: `cpu_pinned_double_buffer + pybind11 + sync learner`
3. B fixed learner-light: `cpu_pinned_double_buffer + pybind11` with learner-light behavior fixed as the B path default

The key conclusion is:

> B fixed learner-light is a memory-efficiency infra improvement, not a claim that B path is faster than A path in pure throughput.

From the report's 1000-iteration comparison:

| Run | E2E s | E2E steps/s | cycle mean ms |
|---|---:|---:|---:|
| A `gpu_cache` | 193.020 | 10610.289 | 189.952 |
| B current `pybind11 sync` | 197.060 | 10392.767 | 193.609 |
| B fixed learner-light | 195.823 | 10458.429 | 192.138 |

Report interpretation:

- B fixed is faster than the old B current baseline.
- A path is still about 1.43% faster in E2E throughput in that 1000-iteration replicate.
- B fixed preserves about 98.57% of A path E2E throughput.
- B fixed reduces persistent replay GPU memory from about 1720 MiB to about 107.5 MiB.

Memory comparison from the report:

| Metric | A `gpu_cache` | B fixed learner-light |
|---|---:|---:|
| Persistent replay GPU memory | ~1720 MiB | ~107.5 MiB |
| Replay GPU memory reduction | baseline | ~93.75% lower |
| Persistent replay-related total | ~3440 MiB | ~1935 MiB |

## Additional Validation on This PR Branch

After migration, this branch was validated against the round16 B fixed learner-light traces.

Generated PR artifacts:

```text
/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_500/perfetto_offpolicy_timeline.json
/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_1000/perfetto_offpolicy_timeline.json
```

Round16 baselines:

```text
/home/fxl/unilab_workspace/round16/artifacts/learner_light_B_pybind11_light_500_r1/perfetto_offpolicy_timeline.json
/home/fxl/unilab_workspace/round16/artifacts/learner_light_B_pybind11_light_1000_r1/perfetto_offpolicy_timeline.json
```

Analyzer results:

| Run | round16 B fixed | PR branch B path |
|---|---:|---:|
| 500 iter `training_e2e` | 96.988 s | 96.396 s |
| 500 iter cycle mean | 187.832 ms | 186.838 ms |
| 1000 iter `training_e2e` | 195.823 s | 185.206 s |
| 1000 iter cycle mean | 192.138 ms | 182.141 ms |

Trace markers were checked on the PR branch:

```text
pipeline=cpu_pinned_double_buffer
pack_layout=packed
pack_executor=collector_thread
h2d_submitter=pybind11
direct_pinned_shared=True
weight sync mode=sync
```

## User-Facing / Training Impact

- SAC training now defaults to B path through `training.replay_pipeline=auto`.
- Existing A path can still be selected explicitly:

```bash
uv run train --algo sac --task g1_walk_flat --sim mujoco \
  training.replay_pipeline=gpu_cache
```

- Default SAC B path command:

```bash
uv run train --algo sac --task g1_walk_flat --sim mujoco
```

- TD3 and FlashSAC remain on `gpu_cache` when `training.replay_pipeline=auto`.
- B path rejects unsupported configurations early:
  - non-SAC algo with `cpu_pinned_double_buffer`;
  - SAC multi-GPU;
  - non-CUDA device;
  - `training.no_sync_collection=true`.

## Validation

- [x] `make test-all` after rebase onto `origin/main` (`dd46ec1`, 2026-05-10): ruff, mypy, pyright, pytest passed (`810 passed, 11 skipped, 241 deselected`); pyright kept the existing `mlx.core` missing-import warning.
- [x] `ruff check .`
- [x] `ruff format --check .`
- [x] `mypy src/unilab`
- [x] `pyright`
- [x] focused off-policy B path tests
- [x] `pytest -m "not slow" --cov=unilab --cov-report=term-missing --cov-fail-under=25`
- [x] SAC B path iter500 performance validation
- [x] SAC B path iter1000 performance validation
- [x] timeline analyzer comparison against round16 B fixed learner-light traces
- [x] trace marker validation

Commands actually run:

```bash
/home/fxl/miniforge3/envs/fxl/bin/uv run train --algo sac --task g1_walk_flat --sim mujoco \
  training.replay_pipeline=cpu_pinned_double_buffer \
  training.no_play=true \
  training.trace_enabled=true \
  training.trace_output_dir=/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_500 \
  algo.max_iterations=500 \
  algo.save_interval=500

/home/fxl/miniforge3/envs/fxl/bin/uv run train --algo sac --task g1_walk_flat --sim mujoco \
  training.replay_pipeline=cpu_pinned_double_buffer \
  training.no_play=true \
  training.trace_enabled=true \
  training.trace_output_dir=/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_1000 \
  algo.max_iterations=1000 \
  algo.save_interval=1000

/home/fxl/miniforge3/envs/fxl/bin/uv run python scripts/analyze_offpolicy_trace.py \
  /home/fxl/unilab_workspace/round16/artifacts/learner_light_B_pybind11_light_500_r1/perfetto_offpolicy_timeline.json \
  /home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_500/perfetto_offpolicy_timeline.json \
  --training-e2e \
  --iteration-resume-gap \
  --event learner/replay_sample \
  --event learner/weight_sync_write \
  --event replay_pipeline/batch_h2d_submit \
  --event gpu/replay_pipeline_batch_h2d

/home/fxl/miniforge3/envs/fxl/bin/uv run python scripts/analyze_offpolicy_trace.py \
  /home/fxl/unilab_workspace/round16/artifacts/learner_light_B_pybind11_light_1000_r1/perfetto_offpolicy_timeline.json \
  /home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_1000/perfetto_offpolicy_timeline.json \
  --training-e2e \
  --iteration-resume-gap \
  --event learner/replay_sample \
  --event learner/weight_sync_write \
  --event replay_pipeline/batch_h2d_submit \
  --event gpu/replay_pipeline_batch_h2d

/home/fxl/miniforge3/envs/fxl/bin/uv run ruff check .
/home/fxl/miniforge3/envs/fxl/bin/uv run ruff format --check .
/home/fxl/miniforge3/envs/fxl/bin/uv run mypy src/unilab
/home/fxl/miniforge3/envs/fxl/bin/uv run pyright

GIT_DIR=/home/fxl/unilab_workspace/main_new/UniLab/.git/worktrees/UniLab \
GIT_WORK_TREE=/home/fxl/unilab_workspace/pr_sac_b_path/UniLab \
COVERAGE_FILE=/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/default_b_path.coverage \
/home/fxl/miniforge3/envs/fxl/bin/uv run pytest -m "not slow" \
  --cov=unilab \
  --cov-report=term-missing \
  --cov-fail-under=25 \
  -o cache_dir=/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/pytest_cache_default_b_path_full
```

Latest test result:

```text
776 passed, 20 skipped, 241 deselected
coverage: 59.34%
```

Pyright result:

```text
0 errors, 1 warning
```

The warning is the existing `mlx.core` missing-import warning in `src/unilab/training/seed.py`.

## Impact

- Backend impact: `mujoco` SAC validation was run; implementation is off-policy runner/replay infra.
- Platform impact: Linux/WSL CUDA path for B path; A path remains available.
- Training effect expected: yes. SAC defaults to B path and should use much less persistent replay GPU memory.
- Docker impact: none expected.
- Public API impact: new `training.replay_pipeline=auto` behavior and explicit `gpu_cache`/`cpu_pinned_double_buffer` selection.

## Artifacts

- PR branch B path 500 trace:
  - `/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_500/perfetto_offpolicy_timeline.json`
- PR branch B path 1000 trace:
  - `/home/fxl/unilab_workspace/pr_sac_b_path/artifacts/b_path_1000/perfetto_offpolicy_timeline.json`
- Round16 report:
  - `/home/fxl/unilab_workspace/report/offpolicy_b_fixed_learner_light_infra_report_zh.md`

## Risks / Notes

- B path is not presented as a pure-speed improvement over A path.
- The strongest argument is memory efficiency: near-A throughput with much lower persistent replay GPU memory.
- B path currently moves more H2D data per learner iteration than A path lazy sync.
- The memory numbers in the report are derived from tensor shapes, trace bytes, and code paths, not from `nvidia-smi` peak measurements.
- Future work should add CUDA allocator counters and explore reducing B path packed width / H2D volume.

## Checklist

- [x] Added or updated tests where needed.
- [x] Preserved explicit A path fallback.
- [x] Kept B path SAC-only guardrails.
- [x] Validated iter500/iter1000 performance before CI.
- [x] Ran local CI-style validation.
- [ ] Link the driving issue, if available.
- [ ] Verify GitHub Actions after pushing the PR branch.

### CODE REVIEW

**Comment by TATP-233** [comment]
Follow-up split from #302 after the off-policy scope moved away from full GPU replay mirror:

- #432 tracks the remaining APPO staging work (`replay_queue` / whole-rollout `torch.cat` / bounded staging pool).
- #433 tracks ROCm real-hardware validation for the replay double-buffer transfer backend.

#302 can now be closed as superseded by #404 for off-policy replay, with the remaining work tracked in the two follow-up issues above.
